### PR TITLE
[CINFRA-498] Replaced explicit-destroy/placement-new hack with optionals

### PR DESCRIPTION
**Pull request by goedderz**
### Scope & Purpose

Use optionals instead of an old explicit destroy & placement-new hack to reset executors and fetchers.

- [X] :hankey: Bugfix

### CODE REVIEW

**Comment by maierlars** [comment]
The remaining changes are in the following tickets: https://arangodb.atlassian.net/browse/CINFRA-505

File: arangod/Aql/ExecutionBlockImpl.cpp [hunk 1]
```diff
@@ -272,8 +272,15 @@ std::unique_ptr<OutputAqlItemRow> ExecutionBlockImpl<Executor>::createOutputRow(
 }
 
 template<class Executor>
-Executor& ExecutionBlockImpl<Executor>::executor() {
-  return _executor;
+auto ExecutionBlockImpl<Executor>::executor() noexcept -> Executor& {
+  TRI_ASSERT(_executor.has_value());
+  return *_executor;
+}
```
Comment (MBkkt) [inline arangod/Aql/ExecutionBlockImpl.cpp:278]:
```suggestion
auto& ExecutionBlockImpl<Executor>::executor() noexcept {
  TRI_ASSERT(_executor.has_value());
  return *_executor;
}
```

File: arangod/Aql/ExecutionBlockImpl.cpp [hunk 2]
```diff
@@ -272,8 +272,15 @@ std::unique_ptr<OutputAqlItemRow> ExecutionBlockImpl<Executor>::createOutputRow(
...
+
+template<class Executor>
+auto ExecutionBlockImpl<Executor>::fetcher() noexcept -> Fetcher& {
+  TRI_ASSERT(_rowFetcher.has_value());
+  return *_rowFetcher;
 }
```
Comment (MBkkt) [inline arangod/Aql/ExecutionBlockImpl.cpp:284]:
```suggestion
auto& ExecutionBlockImpl<Executor>::fetcher() noexcept {
  TRI_ASSERT(_rowFetcher.has_value());
  return *_rowFetcher;
}
```

**Review by MBkkt — APPROVED** [review]

File: arangod/Aql/ExecutionBlockImpl.cpp (hunk 1 above)
Comment (goedderz) [reply arangod/Aql/ExecutionBlockImpl.cpp]: I'd prefer that, too, but it doesn't match the declaration.

**Pull request merged**